ta/pkcs11/src/processing_symm.c

```diff
@@ -712,3 +796,70 @@ enum pkcs11_rc step_symm_operation(struct pkcs11_session *session,
 
 	return rc;
 }
+
+enum pkcs11_rc derive_key_by_symm_enc(struct pkcs11_session *session,
+				      struct obj_attrs **head)
+{
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct active_processing *proc = session->processing;
+	struct input_data_ref *input = proc->extra_ctx;
+	void *in_buf = NULL;
+	void *out_buf = NULL;
+	uint32_t out_size = 0;
+	uint32_t in_size = 0;
+	uint32_t size = sizeof(uint32_t);
+	uint32_t key_length = 0;
+
+	if (!proc->extra_ctx)
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	in_buf = input->data;
+	in_size = input->size;
+
+	out_size = in_size;
+	out_buf = TEE_Malloc(out_size, 0);
+	if (!out_buf)
+		return PKCS11_CKR_DEVICE_MEMORY;
+
+	res = TEE_CipherDoFinal(proc->tee_op_handle, in_buf, in_size, out_buf,
+				&out_size);
+	rc = tee2pkcs_error(res);
+	if (rc)
+		goto out;
+
+	/* Get key size if present in template */
+	rc = get_attribute(*head, PKCS11_CKA_VALUE_LEN, &key_length, &size);
+	if (rc && rc != PKCS11_RV_NOT_FOUND)
+		goto out;
+
+	if (key_length) {
+		/* Derived key is smaller than the required size */
+		if (out_size < key_length) {
+			rc = PKCS11_CKR_DATA_LEN_RANGE;
+			goto out;
+		}
+	} else {
+		key_length = out_size;
+		rc = set_attribute(head, PKCS11_CKA_VALUE_LEN, &key_length,
+				   sizeof(uint32_t));
+		if (rc)
+			goto out;
+	}
+
+	/* Now we can check the VALUE_LEN field */
+	rc = check_created_attrs(*head, NULL);
```

Comment:
This checks VALUE_LEN consistency in the created object.
We should rename it `attrs_check_created_value_len()`. (for the record)

Reply:
Seems like a good idea. I will submit a patch later.